src/Camera/DefaultWorldCameraModel.cpp

```diff
@@ -0,0 +1,784 @@
+	_targetOrigin = _targetFocus + basisZ * _arcBallRadius;
+	_focusAtClick = _targetFocus;
+}
+
+void DefaultWorldCameraModel::UpdateModeDragging(const Camera& camera, glm::u16vec2 mouseCurrent, float mouseMovementDistance)
+{
+	// TODO: Check on optional of _originToHandPlaneNormal to see if there even is a valid grabbing
+
+	// deproject with camera at origin (0, 0, 0), you can use the current
+	// rotation, because during this operation, we cannot rotate so the
+	// current rotation will be the same as rotation at click
+	// Use the current projection matrix too because we don't want fov
+	// changes to affect the current operation.
+	// In fact, using the regular deprojection followed by removing the
+	// current translation will have the same effect
+	const auto depth = glm::mix(0.3f, 3.5f, glm::clamp(camera.GetOrigin().y / 20.0f, 0.0f, 1.0f));
+	const auto mouseDeltaCurrent = ScreenToWorld(camera, mouseCurrent) * depth;
+	const auto mouseDeltaAtClick = ScreenToWorld(camera, _mouseAtClick) * depth;
+
+	// Get dot product of mouse-camera vector and the normal of the up maximizing plane between the camera and hand
+	// Note that mouseAlignment could be unit vectors, but they are scaled by the near plane
+	const auto mouseAlignmentCurrent = glm::dot(mouseDeltaCurrent, _originToHandPlaneNormal);
+	const auto mouseAlignmentAtClick = glm::dot(mouseDeltaAtClick, _originToHandPlaneNormal);
+
+	if ((mouseAlignmentCurrent < -k_HandDragVectorAlignmentThreshold &&
+	     mouseAlignmentAtClick < -k_HandDragVectorAlignmentThreshold) ||
+	    (mouseAlignmentCurrent > k_HandDragVectorAlignmentThreshold &&
+	     mouseAlignmentAtClick > k_HandDragVectorAlignmentThreshold))
+	{
+		const auto originAlignmentAtClick = glm::dot(_originAtClick, _originToHandPlaneNormal);
+		const auto alignmentDelta = _alignmentAtInteractionStart - originAlignmentAtClick;
+
+		const auto alignmentRatioCurrent = alignmentDelta / mouseAlignmentCurrent;
+		const auto alignmentRatioAtClick = alignmentDelta / mouseAlignmentAtClick;
+
+		if (alignmentRatioCurrent > k_HandDragVectorAlignmentRatioThreshold &&
+		    alignmentRatioAtClick > k_HandDragVectorAlignmentRatioThreshold)
+		{
+			auto movement = mouseDeltaCurrent * alignmentRatioCurrent - mouseDeltaAtClick * alignmentRatioAtClick;
+			const auto movementDistance2 = glm::length2(movement);
+			if (movementDistance2 > mouseMovementDistance)
+			{
+				// movement *= mouseMovementDistance *glm::sqrt(movementDistance2); // FIXME: This whole section is broken
+			}
+
+			_targetOrigin = _originAtClick - movement;
+			_targetFocus = _focusAtClick - movement;
+
+			_targetOrigin =
+			    _targetFocus + glm::normalize(_targetOrigin - _targetFocus) * _originFocusDistanceAtInteractionStart;
+		}
+	}
+}
+
+void DefaultWorldCameraModel::UpdateModeFlying(glm::vec3 eulerAngles)
+{
+	if (!_handPosition.has_value() || !_screenSpaceCenterRaycastHit.has_value())
```

Comment:
I tested it and didn't had any more problem with bad optional access.
```suggestion
	if (!_screenSpaceCenterRaycastHit.has_value() || !_screenSpaceMouseRaycastHitAtClick.has_value())
```
It isn't necessary to avoid bad access, but when double clicking on the sky now you fly toward some coordinates behind the camera so i think it would be better to ensure that the mouse hit the land before initiating a fly-by.

Reply:
oh yeah good catch